File: ci/publish/publisher.py
from abc import ABC, abstractmethod
from pathlib import Path
from .models import ReleaseMetadata


class Publisher(ABC):
    def __init__(
        self, artifacts_dir: Path, dry_run: bool = False, verbose: bool = False
    ):
        self.artifacts_dir = artifacts_dir
        self.dry_run = dry_run
        self.verbose = verbose

    @abstractmethod
    def publish(self, release: ReleaseMetadata) -> None:
        pass

File: publish/api/src/main/kotlin/net/xolt/freecam/publish/Publisher.kt
package net.xolt.freecam.publish

import net.xolt.freecam.model.ReleaseMetadata
import net.xolt.freecam.publish.model.GitHubConfig
import java.nio.file.Path

interface Publisher : AutoCloseable {
    suspend fun publish(metadata: ReleaseMetadata)
}

interface PublisherFactory {
    fun create(
        dryRun: Boolean,
        artifactsDir: Path,
        githubConfig: GitHubConfig,
    ): Publisher
}